mlflow/store/sqlalchemy_store.py

```diff
@@ -206,43 +221,33 @@ def log_param(self, run_uuid, param):
             self._get_or_create(SqlParam, run_uuid=run_uuid, key=param.key,
                                 value=param.value)
         except sqlalchemy.exc.IntegrityError:
-            raise MlflowException('changing parameter {} value is not allowed'.format((run_uuid,
-                                                                                      param)),
+            self.session.rollback()
+            old_param, _ = self._get_or_create(SqlParam, run_uuid=run_uuid, key=param.key)
+            raise MlflowException("Changing param value is not allowed. Param with key='{}' was"
+                                  " already logged with value='{}' for run ID='{}. Attempted "
+                                  " logging new value '{}'.".format(param.key, old_param.value,
+                                                                    run_uuid, param.value),
                                   INVALID_PARAMETER_VALUE)
 
-    def set_tag(self, run_uuid, tag):
-        run = self._get_run(run_uuid)
-        self._check_run_is_active(run)
-        new_tag = SqlTag(run_uuid=run_uuid, key=tag.key, value=tag.value)
-        self._save_to_db(new_tag)
-
     def get_param(self, run_uuid, param_name):
-        param = self.session.query(SqlParam).get((param_name, run_uuid))
-
+        param = self.session.query(SqlParam).filter_by(run_uuid=run_uuid, key=param_name).first()
         if param is None:
             raise MlflowException('Param={} does not exist'.format(param_name),
                                   RESOURCE_DOES_NOT_EXIST)
+        return param.to_mlflow_entity()
 
-        return param.value
-
-    def get_metric_history(self, run_uuid, metric_key):
-        metrics = self.session.query(SqlMetric.value).filter_by(run_uuid=run_uuid, key=metric_key)
-        values = []
-        for metric in metrics:
-            values.append(metric.value)
-
-        return values
-
-    def get_metric(self, run_uuid, metric_key):
-        metric = self.session.query(SqlMetric.value).filter_by(
-            run_uuid=run_uuid, key=metric_key
-        ).order_by(sqlalchemy.desc(SqlMetric.timestamp)).first()
+    def set_tag(self, run_uuid, tag):
+        run = self._get_run(run_uuid)
+        self._check_run_is_active(run)
+        new_tag = SqlTag(run_uuid=run_uuid, key=tag.key, value=tag.value)
+        self._save_to_db(new_tag)
 
-        if metric is None:
-            raise MlflowException('Metric={} does not exist'.format(metric_key),
+    def get_tag(self, run_uuid, tag_name):
```

Comment:
Do we even need `get_tag`? That's not part of the abstract_store API.

Reply:
removed